Add BaseScopeManager interface

vmarchaud opened the pull request:
Based on https://github.com/open-telemetry/opentelemetry-js/pull/65

I mostly based the API on the Opencensus one, i will explain the choice i made for each function so we have a starting point for the discussion.

vmarchaud commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+
+/** A key/value object for anything that can be stored in the context */
+export interface ContextStore {
+  [key: string]: unknown;
```
I believe the simpler is the API, the better it is. So instead of having a custom class and get/set method i choose to just expose a object there.

rochdev replied:
I think there is value to have methods so that the internal storage can be changed without impacting the caller.

rochdev replied:
I'm actually not sure why a separate interface is needed either since `get/set` could be directly on the manager.

vmarchaud commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+   * handler).
+   * @param fn A function to which to bind the trace context.
+   */
+  wrapFunction<T>(fn: Func<T>): Func<T>;
```
Pretty much the most important function to propagate context, i re-used the signature of the Census implementation.

rochdev replied:
It should be possible to bind to a specific context. Personal preference also to call this `bind` but I don't have a strong opinion.

vmarchaud replied:
What's the use case for binding to specific context ? I believe this function will be only used to propagate the current context, right ?

rochdev replied:
The default should be the current context, but it should be possible to override this with a specific context. This is usually necessary in auto instrumentation where monkey patching sometimes mean you are not in the expected context which you need to override.

vmarchaud commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+   * There are specific case where the context is not propagated by specific
+   * code (ex: user space queueing) so expect it to be null sometimes.
+   */
+  getCurrentContext(): ContextStore | null;
```
We should also choose if we return a empty object or null when there are no active context. I would prefer to use null there since users would not trust it's content.

draffensperger replied:
I general I dislike `null` when we can avoid it, but I think you have a good point here - there is a semantic distinction between an empty object just because the context is empty and an empty object because there is no context available. I think that distinction is useful to preserve, so I like keeping it.

rochdev replied:
I agree that `null` better clarifies the intent.

rochdev replied:
I generally prefer smaller method names when possible. Any chance we could use something like `current()`? For a Context Manager, I would expect that to return the current context since that's what is managed by the class.

vmarchaud replied:
@rochdev I don't really mind using `current()` but we should be consistent on this subject, do we avoid using get/set prefix or use them everywhere.
Seems that other PR's tend to go without prefix so i will update to match that.

rochdev replied:
I agree we should be consistent. In general I tend to prefer omitting `get` since it's usually redundant. Verbs can be used for actions and nouns for getters. The only downside is for words that can be both, but then we can add `get` if the API would otherwise be unclear.

vmarchaud commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+   * @param emitter An event emitter whose handlers should have
+   *     the trace context binded to them.
+   */
+  wrapEmitter(emitter: NodeJS.EventEmitter): void;
```
I believe the `wrapEmitter` is not needed anymore since we target node 8, `async_hooks` propagate the context correctly with event emitters.

Also it might be a problem to compile to the browser since the EventEmitter will not be available in this environment.

I added it so we can make a decision on its utility

draffensperger replied:
Thanks for thinking of the browser here! If we can remove it, that would be awesome.

When I built OpenCensus Web, I had to do a workaround to avoid having to bring in the Node TS types (which felt like an extra dependency and risks type conflicts). So what I ended up doing was writing a script that did some auto-replacements and I also inlined the type for the NodeJS.EventEmitter.

If we do keep this, I would ask us to consider inlining the event emitter type in this library so that browser users of it don't need the Node TS types to make it compile for TS.

rochdev replied:
`async_hooks` doesn't propagate the correct context, at least to my knowledge. It will propagate the ID where `emit` is run and not where `on` is run.

Also, it would make for a simpler interface to have a single `wrap` method that supports multiple types. It will also allow extensions without breaking the contract.

vmarchaud replied:
Yeah i though of using only one method in the first place but i remove it later thinking it was bringing too much complexity. Thinking it again it might be problematic with the browser API once again, i will change to only one method

rochdev commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+   * code (ex: user space queueing) so expect it to be null sometimes.
+   */
+  getCurrentContext(): ContextStore | null;
+}
```
There should also be a `wrapPromise` similar to `bindEmitter`, but as mentioned above I would merge all of them into a single `wrap` method for extensibility.

rochdev commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+  [key: string]: unknown;
+}
+
+export interface ContextManager {
```
This interface makes it unclear if child contexts should inherit from parent contexts. While that's the behavior of `continuation-local-storage` which is arguably the grandfather of JS context propagation, there is a performance cost of doing so and it's prone to memory leaks.

Not supporting that however makes this interface potentially too flexible, in which case I would probably get/set a single object and not allow individual properties.

rochdev replied:
I thought about this more, and I think think we should go with the second approach where we store a single object. Then there should be a method that runs a callback with the object as its context. This will be much simpler and not ambiguous.

In the future, it will be possible to add an API similar to what is proposed here on top as another project, something like `@opentelemetry/cls` which would allow inheritance and individual properties.

vmarchaud replied:
I don't know much about the performance of each approach but i would prefer to stick on the behavior of the CLS for the sake of simplicity.
Almost all APM vendors use it so it's well known, changing it might make difficult the process of porting a instrumentation from any vendor to the OpenTelemetry API, which i think is more problematic than already-established performance cost (since the CLS approach is currently used anyway).

mayurkale22 replied:
Do we need `enable()` and `disable()` methods or by default it would be enable immediately after creation?

rochdev replied:
> I don't know much about the performance of each approach but i would prefer to stick on the behavior of the CLS for the sake of simplicity.

I did many benchmarks and not using CLS is faster and safer. It's also easier to reason about to store a single object since it ensures different vendors have full control over how they implement this if they need to have their own implementation. What's also great about the simpler approach is that you can wrap it and build a CLS on top. The opposite is not true.

At Datadog we started with CLS and switched later on for performance and stability reasons. I think most vendors went with CLS because it's the common way to propagate context for applications, but it doesn't mean it's the best way for this type of project, it's just the default.

> Do we need enable() and disable() methods or by default it would be enable immediately after creation?

I would say yes. We have customers who wanted more control over context propagation. It also makes it easier to be used on its own and for integration testing.

draffensperger commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+ */
+
+/** Default type for functions */
+// tslint:disable:no-any
```
Can this be removed now that you use `unknown` below?

mayurkale22 commented:
@vmarchaud https://github.com/open-telemetry/opentelemetry-js/pull/65 has been merged, please rebase with the master.

vmarchaud commented:
I've changed the interface based on all feedbacks @mayurkale22 @rochdev @draffensperger

rochdev commented on packages/opentelemetry-context-base/src/types.ts:
```diff
@@ -0,0 +1,51 @@
+ * limitations under the License.
+ */
+
+export interface BaseContext {
```
I still think a simpler approach that doesn't support getting and setting individual properties would serve us better internally. Then we can always extend with a separate package if needed.

Something like:

```ts
export interface BaseContext {
  active: unknown; // the current active context
  activate<T>(object: unknown, fn: T); // activate an object on the current context for the given function
  bind<T>(target: T, object?: unknown): T; // "wrap" is semantically incorrect for objects
  enable(): void; // could also return the instance
  disable(): void;
}
```

I also think using "scope" instead of "context", similar to OpenTracing, would be better given how "context" is already overloaded in so many ways in the project already.

draffensperger replied:
A couple questions on this:
- Would it make sense to have `object` typed as `{[key: string]: unknown}` so that callers could depend on the context object being a map? That could allow us to e.g. store some OpenTelemetry specific context in a given key on it without disrupting other users of the context.
- Could you help me understand the `activate` function? I assume that `bind` creates a new function with hooks around the passed function such that when the new bound function it will invoke `target`  (or the emitted event handler, etc.) such that its current will be set to the current or specified context. Is `activate` saying that it will change the default context for the function in a mutating way? Does it call the function right away?

rochdev replied:
> That could allow us to e.g. store some OpenTelemetry specific context in a given key on it without disrupting other users of the context.

Other users of would use their own `BaseContext` instance so there would not be a conflict.

> Could you help me understand the activate function?

The main difference is `activate()` is meant to just activate a span immediately in the callnack, but `bind()` returns a wrapper that when called will pass its argument to the delegate and activate the span as well.

For example:

```js
scope.activate(span, () => {
  // no arguments are passed
  // scope.active() returns the span in every async context created here
})

const fn = (foo) => {
  // foo='bar'
  // scope.active() returns the span from below in every async context created here
}

const wrapper = scope.bind(fn, span)

wrapper('bar')
```

draffensperger replied:
OK, I see! I think I'm used to that function being called something like `run` or some derivative e.g. `runInAsyncScope`.

vmarchaud replied:
I find the term `activate` on the object not really transparent to it's behavior, what do you think about `setAsCurrent` and `current()` or `setAsActive` and `active` ?

vmarchaud replied:
Also, i agree with @rochdev that the `context` word is already used a lot in the implementation and using the `scope` word is not wrong, so i'm in favor of using it

rochdev replied:
You are not really setting it as current though. You are activating the span on the scope of the function that is passed as a parameter. I'm not sure if it's 100% clear, but it's definitely semantically correct.

Zones use the work `fork` for this. CLS uses `run`. The spec uses `withSpan` but on the tracer.

My personal preference would be to find any single word that correctly represents what this does, which I think `activate` does but I may be biased as I've been using this terminology for a long time from [OpenTracing](https://github.com/opentracing/specification/blob/master/rfc/scope_manager.md#scopemanager) at this point.

vmarchaud replied:
I would personally prefer to use `with` since it's in the spec and clearer than `active` to me

rochdev replied:
So `with()` and `current()`? Do you think `bind()` makes sense as well?

vmarchaud replied:
Totally fine with `bind`

vmarchaud commented:
Addressed comments by @rochdev. I've also renamed the packages to `scope-base` following comment by @rochdev saying that `Context` was already heavily used across the implementation so using `scope` will be clearer.

mayurkale22 commented on packages/opentelemetry-scope-base/package.json:
```diff
@@ -1,13 +1,14 @@
   "scripts": {
-    "test": "nyc ts-mocha -p tsconfig.json test/**/*.ts",
-    "tdd": "yarn test -- --watch-extensions ts --watch",
+    "test": "nyc ts-mocha -p ./tsconfig.json test/**/*.ts",
```
I think `test`, `tdd` and `codecov` scripts is not required for this package (atleast for now), unless we are adding tests for `BaseScopeManager`.

rochdev replied:
I don't think there is any harm to keep them since the implementation of the base should be coming right after this PR.

draffensperger commented on packages/opentelemetry-scope-base/src/types.ts:
```diff
@@ -0,0 +1,46 @@
+   * @param scope Any object to set as the current active scope
+   * @param fn A callback to be imediately run within a specific scope
+   */
+  with<T>(scope: unknown, fn: T): void;
```
Could you specify that the type for `fn` must be a function?

E.g.: `with(scope: unknown, fn: (...args: unknown[]) => unknown): void`?

Should `with` return what the executed function returns?

vmarchaud replied:
I believe we could just use `with(scope: unknown, fn: Function): void` right ? I don't see the point of typing the args there

draffensperger commented:
Given that this is just an interface for now, would it make sense to have it in `@opentelemetry/types`? What's the motivation for a separate package for the base scope?

I like the idea of having the actual implementations be in their own packages, but not sure the interface needs its own. @rochdev WDYT?

vmarchaud commented:
@draffensperger I believe the goal was the package to be usable outside of opentelemetry, so having the interface in `core` would mean add another dependency.

mayurkale22 commented on packages/opentelemetry-scope-base/src/types.ts:
```diff
@@ -0,0 +1,46 @@
+  /**
+   * Enable scope management
+   */
+  enable(): void;
```
Does it make sense to return `this` instead of `void`?

rochdev commented:
LGTM

@mayurkale22 @danielkhan is it fine to use `Scope` instead of `Context` from the spec perspective?

mayurkale22 commented:
> @mayurkale22 @danielkhan is it fine to use `Scope` instead of `Context` from the spec perspective?

I think it is fine to use `Scope`, looks like Java and .NET uses the same terminology https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/context/Scope.java

vmarchaud commented:
Addressed comments once again, i believe it's good to land now.
(@rochdev could you remove the requested changes ? i think it will block the PR otherwise)

draffensperger commented on packages/opentelemetry-scope-base/src/types.ts:
```diff
@@ -0,0 +1,46 @@
+   * @param scope Any object to set as the current active scope
+   * @param fn A callback to be imediately run within a specific scope
+   */
+  with(scope: unknown, fn: Function): void;
```
Should this return the return type of the function that is called?

hekike replied:
is the coupling between https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-types/src/trace/tracer.ts#L72

```
// in tracer: 
withSpan(span, fn) {
  return this.scopeManager.with(span, function() {
    return fn();
 });
}
```

// in userland:
```
server.use(function(req, res, next) {
     const span = tracer.starr('server');
     tracer.withSpan(span, function() {
       next();
     });
});
```

or do I miss something?

vmarchaud replied:
@draffensperger Yes, i've updated it.
@hekike Yes, you could however simplify with :
```
withSpan(span, fn) {
  return this.scopeManager.with(span, fn);
}
```
and
```
server.use(function(req, res, next) {
  const span = tracer.start('server');
  tracer.withSpan(span, next);
});
```

mayurkale22 commented:
> I would rename the first parameter of `bind` to `target` since it may not be an object but otherwise LGTM.

@vmarchaud Feel free to handle it in separate PR, merging it now. Thanks for the work and PR!
